## Importamos librerias

In [1]:
import pandas as pd
import numpy as np

# Librerias de visualización
import matplotlib.pyplot as plt
import seaborn as sns

## Cargamos archivo

In [2]:

df_bank = pd.read_csv("../data/output/bank-additional.csv")
pd.set_option('display.max_columns', None)
df_bank.sample(5)

Unnamed: 0   age          job   marital          education  default  \
3667         3764  37.0   technician   MARRIED        high.school      0.0   
40051       40218  82.0    housemaid  DIVORCED           basic.4y      0.0   
14187       14284  35.0   technician    SINGLE  university.degree      0.0   
3336         3433  36.0  blue-collar   MARRIED           basic.4y      NaN   
3097         3194  37.0   technician    SINGLE  university.degree      0.0   

       housing  loan    contact  duration  campaign  pdays  previous  \
3667       1.0   0.0  telephone       136         1    999         0   
40051      0.0   0.0   cellular       316         1    999         0   
14187      1.0   0.0   cellular       118         4    999         0   
3336       1.0   0.0  telephone       328         4    999         0   
3097       1.0   0.0  telephone        30         1    999         0   

          poutcome  emp.var.rate cons.price.idx cons.conf.idx euribor3m  \
3667   NONEXISTENT           1.1         93,994         -36,4     4,859   
40051  NONEXISTENT          -1.7         94,215         -40,3     0,881   
14187  NONEXISTENT           1.4         93,918         -42,7     4,962   
3336   NONEXISTENT           1.1         93,994         -36,4      4,86   
3097   NONEXISTENT           1.1         93,994         -36,4       NaN   

      nr.employed    y               date  latitude  longitude  \
3667         5191   no  8-septiembre-2017    43.134   -110.889   
40051      4991,6  yes      22-abril-2015    34.425    -76.037   
14187      5228,1   no   4-diciembre-2018    46.965   -120.893   
3336         5191   no  12-diciembre-2016    47.635    -75.807   
3097         5191   no  15-noviembre-2015    33.234    -69.378   

                                        id_  
3667   cdbc80e1-21ef-4a5b-ad28-a2399c97bdbf  
40051  dc3e9f27-fd0f-4b85-88d1-79a75165688e  
14187  68113d2c-d0fc-45f0-ba79-5eb1e84898c9  
3336   9321118e-a67f-497f-a5cf-b9546d31f808  
3097   e1d6e7c3-5cc7-46f7-b8b6-affcd37a3ad2

## Revisamos la informacion del DF

In [3]:
df_bank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      43000 non-null  int64  
 1   age             37880 non-null  float64
 2   job             42655 non-null  object 
 3   marital         42915 non-null  object 
 4   education       41193 non-null  object 
 5   default         34019 non-null  float64
 6   housing         41974 non-null  float64
 7   loan            41974 non-null  float64
 8   contact         43000 non-null  object 
 9   duration        43000 non-null  int64  
 10  campaign        43000 non-null  int64  
 11  pdays           43000 non-null  int64  
 12  previous        43000 non-null  int64  
 13  poutcome        43000 non-null  object 
 14  emp.var.rate    43000 non-null  float64
 15  cons.price.idx  42529 non-null  object 
 16  cons.conf.idx   43000 non-null  object 
 17  euribor3m       33744 non-null 

In [4]:
df_bank.shape

(43000, 24)

In [5]:
df_bank.describe().T

count          mean           std      min          25%  \
Unnamed: 0    43000.0  20651.099047  11868.770913    0.000  10381.75000   
age           37880.0     39.977112     10.437957   17.000     32.00000   
default       34019.0      0.000088      0.009390    0.000      0.00000   
housing       41974.0      0.535998      0.498708    0.000      0.00000   
loan          41974.0      0.155620      0.362499    0.000      0.00000   
duration      43000.0    257.739279    258.666033    0.000    102.00000   
campaign      43000.0      2.567233      2.772294    1.000      1.00000   
pdays         43000.0    962.330953    187.260394    0.000    999.00000   
previous      43000.0      0.174023      0.497366    0.000      0.00000   
emp.var.rate  43000.0      0.077128      1.573898   -3.400     -1.80000   
latitude      43000.0     36.856697      7.225948   24.396     30.61475   
longitude     43000.0    -95.939067     16.752282 -124.997   -110.49425   

                     50%          75%        max  
Unnamed: 0    20642.5000  30930.25000  41187.000  
age              38.0000     47.00000     98.000  
default           0.0000      0.00000      1.000  
housing           1.0000      1.00000      1.000  
loan              0.0000      0.00000      1.000  
duration        179.0000    319.00000   4918.000  
campaign          2.0000      3.00000     56.000  
pdays           999.0000    999.00000    999.000  
previous          0.0000      0.00000      7.000  
emp.var.rate      1.1000      1.40000      1.400  
latitude         36.7610     43.11325     49.384  
longitude       -95.8995    -81.42775    -66.937

Del método describe podemos destacar:
- *Columna "pdays"* Vemos que en el mayor de los casos el resultado es 999, intuimos que este número es erroneo, entendiendo que este valor puede significar que no se ha contactado recientemente con la mayoria de los clientes. Dado que los valores de esta columna expresas *Número de días que han pasado desde la última vez que se contactó con el cliente durante esta campaña.* Entendemos 0 como que aun no ha pasado un dia desde el ultimo contacto. Necesitamos ver la distribucion de los valores
- Para los valores booleanos de las columnas 
   -  *default* - entenemos que a pesar de que hay valores nulos hay mayor porcentaje de valores = 0
   -  *housing*- entenemos que a pesar de que hay valores nulos hay mayor porcentaje de valores = 1 
   -  *loan* - entenemos que a pesar de que hay valores nulos hay mayor porcentaje de valores = 0
   -  Es decir, en base a los datos actuales que tenemos podemos decir que la mayoria de clientes tienen prestamo hipotecario y no tienen otro tipo de préstamo, así como tampoco impagos. Sin embargo, tenemos que concretar aun el porcentaje de nulos que tenemos para poder realizar una conclusion al respecto.
- En la columna *campaing* parece que hay mucha diferencia entre la media/mediana y el valor máximo por lo que puede que haya algun valor atípico en el extremo.

Dado que hay columnas aun con "," en los números y las identifica con valores de tipo "object" vamos a hacer este cambio para poder ver todos los datos.

In [6]:
df_bank.isna().sum().sort_values(ascending=False)

euribor3m         9256
default           8981
age               5120
education         1807
loan              1026
housing           1026
cons.price.idx     471
job                345
date               248
marital             85
Unnamed: 0           0
contact              0
pdays                0
previous             0
campaign             0
duration             0
emp.var.rate         0
poutcome             0
nr.employed          0
cons.conf.idx        0
y                    0
latitude             0
longitude            0
id_                  0
dtype: int64

In [7]:
porcentaje_nulos = df_bank.isna().mean()*100
porcentaje_nulos.sort_values(ascending=False)

euribor3m         21.525581
default           20.886047
age               11.906977
education          4.202326
loan               2.386047
housing            2.386047
cons.price.idx     1.095349
job                0.802326
date               0.576744
marital            0.197674
Unnamed: 0         0.000000
contact            0.000000
pdays              0.000000
previous           0.000000
campaign           0.000000
duration           0.000000
emp.var.rate       0.000000
poutcome           0.000000
nr.employed        0.000000
cons.conf.idx      0.000000
y                  0.000000
latitude           0.000000
longitude          0.000000
id_                0.000000
dtype: float64

Las columnas con una mayor cantidad de nulos son *euribor3m* y *default* con un porcentaje moderado, *age* con un porcentaje moderado bajo, y las columnas *education*  4.20%
*loan* 2.38 %
*housing* 2.38 %
*cons.price.idx* 1.09 %
*job* 0.80 %
*date*  0.57 %
*marital* 0.19% con porcentaje bajo de nulos


### Comprobamos si hay duplicados

In [8]:
df_bank.duplicated().sum()

np.int64(0)

In [9]:
df_bank["pdays"].value_counts()

pdays
999    41412
3        457
6        428
4        121
9         67
7         66
2         63
12        61
10        55
5         50
13        38
11        30
15        26
1         26
14        22
8         20
0         15
16        13
17        10
18         7
22         4
19         3
21         2
25         1
26         1
27         1
20         1
Name: count, dtype: int64

# Limpieza general de los datos

### Ponemos todos los valores de las columnas "object" en minúsculas

In [10]:
for col in df_bank.select_dtypes(include="O").columns:
    df_bank[col] = df_bank[col].str.lower()

df_bank.tail(5)

Unnamed: 0   age          job  marital            education  default  \
42995       19154   NaN       admin.  married    university.degree      0.0   
42996       26206  34.0   technician  married  professional.course      0.0   
42997       15046   NaN  blue-collar   single             basic.6y      0.0   
42998       15280   NaN       admin.  married    university.degree      NaN   
42999       27570   NaN   unemployed   single    university.degree      0.0   

       housing  loan   contact  duration  campaign  pdays  previous  \
42995      0.0   0.0  cellular       618         2    999         0   
42996      1.0   1.0  cellular        42         7    999         0   
42997      1.0   0.0  cellular       391         2    999         0   
42998      0.0   0.0  cellular       674         3    999         0   
42999      0.0   1.0  cellular       104         2    999         0   

          poutcome  emp.var.rate cons.price.idx cons.conf.idx euribor3m  \
42995  nonexistent           1.4         93,444         -36,1       NaN   
42996  nonexistent          -0.1           93,2           -42       NaN   
42997  nonexistent           1.4         93,918         -42,7       NaN   
42998  nonexistent           1.4         93,918         -42,7     4,958   
42999  nonexistent          -0.1           93,2           -42     4,021   

      nr.employed    y                date  latitude  longitude  \
42995      5228,1  yes     13-octubre-2015    38.147   -105.582   
42996      5195,8   no       17-marzo-2018    49.235   -112.201   
42997      5228,1   no  15-septiembre-2016    40.679   -120.015   
42998      5228,1   no  23-septiembre-2019    27.772   -117.518   
42999      5195,8   no    6-noviembre-2019    41.146   -105.026   

                                        id_  
42995  4eed05de-2a98-4227-b488-32122009b638  
42996  0f0aca88-4088-4fe2-905f-44fb675d9493  
42997  cadadd4b-7ee5-4019-b13a-ca01bb67ca5b  
42998  5f432048-d515-4bb5-9c94-62db451f88d4  
42999  993bbbd6-4dbc-4a40-a408-f91f8462bee6

### Cambiamos todas las "," por "." que hay en las columnas categorizadas como "object" y las dejamos en tipo "float"

In [11]:
def comas (df):
    for col in df.select_dtypes(include="O"):

        df[col] = df[col].str.replace(",", ".")
        try:
            df[col] = df[col].astype("float64")
        except:
            pass
comas (df_bank)
df_bank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      43000 non-null  int64  
 1   age             37880 non-null  float64
 2   job             42655 non-null  object 
 3   marital         42915 non-null  object 
 4   education       41193 non-null  object 
 5   default         34019 non-null  float64
 6   housing         41974 non-null  float64
 7   loan            41974 non-null  float64
 8   contact         43000 non-null  object 
 9   duration        43000 non-null  int64  
 10  campaign        43000 non-null  int64  
 11  pdays           43000 non-null  int64  
 12  previous        43000 non-null  int64  
 13  poutcome        43000 non-null  object 
 14  emp.var.rate    43000 non-null  float64
 15  cons.price.idx  42529 non-null  float64
 16  cons.conf.idx   43000 non-null  float64
 17  euribor3m       33744 non-null 

In [12]:
df = df_bank.copy()

### Modificamos la columna *date* a tipo fecha

In [13]:
meses = {
    "enero": "01",
    "febrero": "02",
    "marzo": "03",
    "abril": "04",
    "mayo": "05",
    "junio": "06",
    "julio": "07",
    "agosto": "08",
    "septiembre": "09",
    "octubre": "10",
    "noviembre": "11",
    "diciembre": "12"
}

df['date'] = df['date'].str.lower()
df['date'] = df['date'].replace(meses, regex=True)

In [14]:
df["date"]= pd.to_datetime(df["date"], format="%d-%m-%Y", errors='coerce')

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Unnamed: 0      43000 non-null  int64         
 1   age             37880 non-null  float64       
 2   job             42655 non-null  object        
 3   marital         42915 non-null  object        
 4   education       41193 non-null  object        
 5   default         34019 non-null  float64       
 6   housing         41974 non-null  float64       
 7   loan            41974 non-null  float64       
 8   contact         43000 non-null  object        
 9   duration        43000 non-null  int64         
 10  campaign        43000 non-null  int64         
 11  pdays           43000 non-null  int64         
 12  previous        43000 non-null  int64         
 13  poutcome        43000 non-null  object        
 14  emp.var.rate    43000 non-null  float64       
 15  co

In [16]:
df.sample(3)

Unnamed: 0   age          job  marital          education  default  \
23086       23183  50.0  blue-collar  married           basic.4y      0.0   
25863       25985  50.0   technician  married  university.degree      0.0   
40443       40610  30.0       admin.   single        high.school      0.0   

       housing  loan    contact  duration  campaign  pdays  previous  \
23086      0.0   0.0  telephone        39         3    999         0   
25863      1.0   0.0   cellular       236         1    999         0   
40443      0.0   0.0   cellular       219         3      6         1   

          poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
23086  nonexistent           1.4          93.444          -36.1      4.965   
25863  nonexistent          -0.1             NaN          -42.0      4.120   
40443      success          -1.1          94.199          -37.5      0.883   

       nr.employed    y       date  latitude  longitude  \
23086       5228.1   no 2016-11-29    40.864   -120.951   
25863       5195.8   no 2018-09-30    32.161   -109.923   
40443       4963.6  yes 2017-01-16    33.522   -113.885   

                                        id_  
23086  a74f52b4-22c9-4c81-aa09-b86990de2cfe  
25863  a55cea13-586a-4265-9d20-80963602b19a  
40443  99bc1e54-3a9a-4edc-9d28-613b8eaf3de2

Los valores de la columna "date" estan en AAA-MM-DD

Aún queda por modificar los tipos de columnas *default* ,*housing* y *loan* de tipo float a int pero una vez que se hayan limpiado de valores nulos

Ahora vamos a ver la distribución por categorias de cada columna categorica

In [17]:
for col in df.select_dtypes(include="object").columns:
        print(col.upper())
        print(df[col].value_counts())    


JOB
job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
Name: count, dtype: int64
MARITAL
marital
married     25999
single      12105
divorced     4811
Name: count, dtype: int64
EDUCATION
education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4y                4356
basic.6y                2386
illiterate                18
Name: count, dtype: int64
CONTACT
contact
cellular     27396
telephone    15604
Name: count, dtype: int64
POUTCOME
poutcome
nonexistent    37103
failure         4461
success         1436
Name: count, dtype: int64
Y
y
no     38156
yes     4844
Name: count, dtype: int64
ID_
id_
089b39d8-e4d0-461b-87d4-814d71e0e079    1
e9d37224-cb6f-4942-98d7-46672963d097    1
3f9f49b5-e410-4948-bf6e-f9244f049

De esta distribución podemos hacer la siguiente explicación de los datos:
   - La mayoria de los clientes tienen la actividad profesional "admin" o "blue-collar"
   - Están casados
   - Tienen estudios universitarios o de instituto
   - El contacto mayor fue por movil
   - No ha habido previo contacto por campaña de marketing
   - No han suscrito producto/servicio
   
Teniendo en cuenta que hay valores nulos que pueden modificar esta conclusion

In [18]:
df.to_csv("../data/bank-cleaned.csv", index=False)